lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -1060,6 +1086,80 @@ def _restart_opensearch(self, event: _RestartOpenSearch) -> None:
+        ):
+            # Also consider peer-relation units
+            hosts.extend([node.ip for node in peer_cm_rel_data.cm_nodes])
+
+        for attempt in Retrying(stop=stop_after_delay(300), wait=wait_fixed(10)):
+            with attempt:
+                nodes = ClusterTopology.nodes(self.opensearch, use_localhost=True, hosts=hosts)
+
+                node = None
+                for node in nodes:
+                    if node.name == self.unit_name:
+                        break
+                if unit_is_stopping and node:
+                    # This is a stop operation, we exclude the unit that is stopping
+                    # if still in the list
+                    nodes.remove(node)
+                    # we can finish the for loop here
+                    break
+                elif not unit_is_stopping and not node:
+                    # This unit is starting, we must assure it shows up in the list
+                    # Let's retry
+                    raise OpenSearchStartError("Node not in the list of nodes")
+                # Any other case is okay to move forward without changes
+                break
+
+        cms = ClusterTopology.get_cluster_managers_names(nodes)
+        # For the sake of predictability, we always sort the cluster managers
+        sorted_cm = sorted(cms)
+        # We always clean the voting.
+        # Then, we deal with the specific cases.
+        self.opensearch_exclusions.delete_voting()
+
+        # Each of the possible cases
+        if len(cms) == 1:
+            # This condition only happens IF the cluster had 2x units and now is scalig down to 1
+            # In this scenario, we should exclude the node that is going away, to force election
+            # to the remaining unit.
+            if unit_is_stopping:
+                self.opensearch_exclusions.add_voting(hosts, node_names=[self.unit_name])
+        elif len(cms) == 2:
+            if unit_is_stopping:
+                # Remove both this unit and the first sorted_cm from the voting
+                self.opensearch_exclusions.add_voting(
+                    hosts, node_names=[self.unit_name, sorted_cm[0]]
+                )
+            else:
+                # We are adding this unit to the cluster and we've waited until it is present
+                # We only exclude one unit:
+                self.opensearch_exclusions.add_voting(hosts, node_names=[sorted_cm[0]])
+            # Now, we clean up the sorted_cm list, as we want to be sure the new manager is elected
+            # and different than the excluded units.
+            sorted_cm.pop(0)
+            # We do not exclude the self.unit_name
+        else:
+            # In this case, we either are scaling down to 0 or len(cms) > 2.
+            # There is nothing more to do then cleanup the exclusions
+            return
+
+        # Now, we must be sure a new manager is elected, or there was a failure
```

Comment:
I'm not sure this is needed / will work. If no CM is elected, we will never have online nodes / API to consume.

Reply:
The node is marked as leaving after 30s, max. The election, if needed, will take place right after. Given we are always excluding the unit that is leaving; if that unit was the elected manager, than it will trigger a re-election right away.

Reply:
This logic has changed following `settle_voting` latest updates.